src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/PinnedGCHandle.T.cs

```diff
@@ -0,0 +1,122 @@
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Represents an strongly-typed opaque, GC handle to a managed object.
+    /// A GC handle is used when an object reference must be reachable from
+    /// unmanaged memory.
+    /// The object is pinned at fixed location in GC heap and allows its
+    /// address to be taken.
+    /// </summary>
+    /// <remarks>
+    /// <see cref="PinnedGCHandle{T}"/> corresponds to Pinned roots.
+    /// For Normal, see <see cref="GCHandle{T}"/>.
+    /// For Weak and WeakTrackResurrection, see <see cref="WeakGCHandle{T}"/>.
+    /// </remarks>
+    /// <seealso cref="GCHandle" />
+    /// <typeparam name="T">The type of the object this <see cref="GCHandle{T}"/> tracks to.</typeparam>
+    public struct PinnedGCHandle<T> : IDisposable
+        where T : class?
+    {
+        // The actual integer handle value that the EE uses internally.
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Allocates a handle for the specified object.
+        /// </summary>
+        /// <param name="target">The object that uses the <see cref="GCHandle{T}"/>.</param>
+        public PinnedGCHandle(T target)
+        {
+            // Unlike GCHandle, pinning any object is allowed
+            _handle = GCHandle.InternalAlloc(target, GCHandleType.Pinned);
+        }
+
+        private PinnedGCHandle(IntPtr handle) => _handle = handle;
+
+        /// <summary>Determine whether this handle has been allocated or not.</summary>
+        public readonly bool IsAllocated => _handle != IntPtr.Zero;
+
+        /// <summary>Gets or sets the object this handle represents.</summary>
+        public readonly T Target
+        {
+            get
+            {
+                IntPtr handle = _handle;
+
+                // Check if the handle was never initialized or was freed.
+                if (handle == IntPtr.Zero)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_HandleIsNotInitialized();
+                }
+
+                // Skip the type check to provide lowest overhead.
+                return Unsafe.As<T>(GCHandle.InternalGet(handle));
+            }
+            set
+            {
+                IntPtr handle = _handle;
+                GCHandle.ThrowIfInvalid(handle);
+
+                // Unlike GCHandle, pinning any object is allowed
+                GCHandle.InternalSet(handle, value);
+            }
+        }
+
+        /// <summary>
+        /// Retrieves the address of object data in a <see cref="PinnedGCHandle{T}"/>.
+        /// </summary>
+        /// <returns>The address of the pinned data object.</returns>
+        [CLSCompliant(false)]
+        public readonly unsafe void* GetAddressOfObjectData()
+        {
+            IntPtr handle = _handle;
+            GCHandle.ThrowIfInvalid(handle);
+
+            return GCHandle.AddrOfPinnedObjectFromHandle(_handle);
```

Comment:
Is this expected to follow GCHandle in special casing strings and arrays here?

Reply:
`Memory<T>.Pin` and `fixed` are all following the contract to return the address of first element. Not following the contract would require strong justification.

Reply:
This method should be only used for non-array non-string blittable types. `Memory<T>.Pin` and `fixed` do not work for non-array non-string blittable types.

I think it would make sense to make it work just for the non-array non-string blittable types and document the behavior as undefined to avoid unnecessary overhead. The overall goal with these new GCHandle types was to eliminate unnecessary overheads.

The original intention with this method was to provide a replacement for GCHandle.AddrOfPinnedObject. I think the method has different enough name to have different behavior.